opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -566,24 +566,112 @@ namespace Opm
     {
-        const std::string msg = std::string("Well potential calculation is not supported for multisegment wells \n")
-                + "A well potential of zero is returned for output purposes. \n"
-                + "If you need well potential to set the guide rate for group controled wells \n"
-                + "you will have to change the " + name() + " well to a standard well \n";
-
-        deferred_logger.warning("WELL_POTENTIAL_NOT_IMPLEMENTED_FOR_MULTISEG_WELLS", msg);
-
         const int np = number_of_phases_;
         well_potentials.resize(np, 0.0);
 
+        updatePrimaryVariables(well_state, deferred_logger);
+
+        // initialize the primary variables in Evaluation, which is used in computePerfRate for computeWellPotentials
+        // TODO: for computeWellPotentials, no derivative is required actually
+        initPrimaryVariablesEvaluation();
+
+        // get the bhp value based on the bhp constraints
+        const double bhp = Base::mostStrictBhpFromBhpLimits(deferred_logger);
+
+        // does the well have a THP related constraint?
+        if ( !Base::wellHasTHPConstraints() ) {
+            assert(std::abs(bhp) != std::numeric_limits<double>::max());
+
+            computeWellRatesWithBhp(ebosSimulator, bhp, /*iterate=*/ true, well_potentials, deferred_logger);
+        } else {
+
+            const std::string msg = std::string("Well potential calculation is not supported for thp controlled multisegment wells \n")
+                    + "A well potential of zero is returned for output purposes. \n"
+                    + "If you need well potential computed from thp to set the guide rate for group controled wells \n"
+                    + "you will have to change the " + name() + " well to a standard well \n";
+
+            deferred_logger.warning("WELL_POTENTIAL_FOR_THP_NOT_IMPLEMENTED_FOR_MULTISEG_WELLS", msg);
+            return;
+        }
+
     }
 
 
+    template<typename TypeTag>
+    void
+    MultisegmentWell<TypeTag>::
+    computeWellRatesWithBhp(const Simulator& ebosSimulator,
+                            const double& bhp,
+                            const bool iterate,
+                            std::vector<double>& well_flux,
+                            Opm::DeferredLogger& deferred_logger)
+    {
+        // store a copy of the well state, we don't want to update the real well state
+        WellState copy = ebosSimulator.problem().wellModel().wellState();
+
+        // try to improve the initial condition.
+        // we adjust the initial seg pressure with half the difference
+        // between the real bhp and the potential bhp.
+        // This is ad-hoc, but seems to work reasonbly well.
+        const double adjustBhp = (bhp - primary_variables_[0][SPres]);
+        double adjustTotalRate = 0.0;
+        for (int p = 0; p < number_of_phases_; ++p) {
+            adjustTotalRate += copy.wellPotentials()[index_of_well_ * number_of_phases_ + p];
```

Comment:
It is easy to understand how the BHP and segment pressure is adjusted. It is hard to understand how the rate is adjusted. Because the primary variable GTotal is a weighted sum of the well rates. When gas exist, this way of scaling is probably highly problematic. 

And what happens if the wellPotentials from the `copy` is not updated yet? then what happens? Please check whether the update of rates is necessary here. 

and also, since in the function, it is using `wellPoentials()` in a explicit way, maybe it is good to change the function name to be `computeWellPotentialsWithBhpLimit`, since it is not a general way of calculating well rates anymore.

Reply:
> Please check whether the update of rates is necessary here.

I think it can be removed. It improved the convergence but only slightly.  I can test without it on some of the big models.